_posts/2016-02-07-coupling-in-distributed-systems.md

```diff
@@ -0,0 +1,88 @@
+
+Second corrective measure would be not simply swallowing those crashes but being able to communicate them to proper actors (being humans or software). Good practice is to provide dead letter queues for poisoned messages just in case we want to manipulate and retry the processing of those messages at that level.
+
+Before of getting into explicit schemas I would say that those measure are not usually enough but they are a good safety net, as shit happens, and we need to be prepared.
+
+### Explicit schemas
+
+If we want to avoid poisoned messages to get into our topics we could provide a middle-man service to intercept and validate explicit schemas. That's an integration service that could be a single point of failure but, at the same time, it could valuable to have a centralised repo of schemas when we have a lot of consumers and the complexity of the system would be hard to grasp in de-centralised fashion.
+
+[Schema registry](http://docs.confluent.io/1.0/schema-registry/docs/index.html) is an example of that for Kafka, and its documentation is full of insights about how to implement that in a distributed, highly available and scalable way.
+
+## Compile approaches
+
+Last approach would be creating software that makes impossible to create messages that not hold the expected schema. Assuming [Scala](http://www.scala-lang.org/), we could create a jar that contains [case classes](http://docs.scala-lang.org/tutorials/tour/case-classes.html) that are object materialisations of a schema.
+
+What are the benefits of this approach?
+
+1. Fail early. We don't have to wait until testing or production to verify that the messages published by some producer are correct.
+2. Centralised knowledge.
+
+What is the solvable problem?
+
+- Cascade updates. If our microservices live in different repos, then we need to make sure that updates into that common binary are applied into producer and consumers. That's cumbersome and if it's not done could generate non expected bugs as we introduced a false sense of security with that library. That could be solved using a [monorepo](http://danluu.com/monorepo/) 
+
+What is the biggest problem?
+
+- Breaking isolation of deployables. One of the points of microservices is being able to deploy its services independently. If you're forced to redeploy N consumer services every time you upgrade the consumer with a non-backward compatible change of the schema library then you're losing that perk. Being able to do small releases is a big enabler of Continuous Delivery, so it's a remarkable loss.
+
+You could argue that only non-backward compatible changes forces a redeploy of consumers and that we should design our schemas in a way that predicts and minimises those kind of changes. 
```

Comment:
Again, this issue already exists if you have an API for which contracts are enforced loosely by implicit schemas and trust. In any case, you'll need versioning to protect your consumers from breaking changes.